include/onnxruntime/core/framework/stream_handles.h

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+#pragma once
+
+#include <functional>
+#include <unordered_map>
+#include "core/framework/ortdevice.h"
+#include "core/common/status.h"
+
+namespace onnxruntime {
+class IExecutionProvider;
+// this opaque handle could be anything the target device generated.
+// it could be a cuda event, or a npu notification implementation
+using NotificationHandle = void*;
+// it can be either a cuda stream, or even nullptr for device doesn't have stream support like cpu.
+using StreamHandle = void*;
+
+namespace synchronize {
+class Notification;
+}
+
+// a stream abstraction which hold an opaque handle, and a reference to which OrtDevice instance this stream belong to.
+// it need to be OrtDevice instance as we might have different stream on different OrtDevice with same type.
+// i.e. different cuda stream on different GPU.
+class Stream {
+ public:
+  Stream(StreamHandle h, const OrtDevice& d) : handle_(h), device_(d) {}
+
+  virtual ~Stream() = default;
+  virtual std::unique_ptr<synchronize::Notification> CreateNotification(size_t /*num_consumers*/) {
+    return {};
+  };
+  // block the host thread until all the tasks in the stream finished.
+  virtual void Flush(){};
+  // The framework may reuse the stream instance for multiple iterations.
+  // This is the API that provide a chance to let the device stream cleanup
+  // resource at the end of a iteration.
+  virtual Status CleanUpOnRunEnd() { return Status::OK(); };
+
+  StreamHandle GetHandle() const { return handle_; }
+
+  const OrtDevice& GetDevice() const { return device_; }
+
+  // We use the timestamp based vector clocks to optimize the resource sharing
+  // between different streams.
+  // Each stream maintain following data structure:
+  // 1. Current timestamp
+  // 2. A lookup table that for a given stream, what is its timestamp when the
+  //    last synchronization happened with current stream.
+  // 3. When a notificaiton is activated, it take a snapshot of current stream's
+  //    lookup table.
+  // 4. When synchronization happened (current stream wait on a notification),
+  //    update its lookup table with the table snapshot in notificaiton.
+  // The memory reusing strategy is:
+  // A kernel in current stream is safe to reuse another stream's memory chunk
+  // as long as the reused chunk's timestamp is less than the last synchonized
+  // timestamp recorded in the lookup table.
+
+  // Get the current timestamp
+  uint64_t GetCurrentTimestamp() const { return timestamp_; }
+
+  // return the timestamp when the last synchronization happened between target stream and current stream.
+  // return 0 if no synchonization happened.
+  uint64_t GetLastSyncTimestampWithTargetStream(Stream* target_stream) const {
```

Comment:
it actually not required. for EP that doesn't support Stream, we will leave the stream as nullptr to represent it. so it could be nullptr. i will document it in the comments.

Reply:
Perhaps, we make want to check for null, which would make it obvious and avoid lookup.

Reply:
yes i made the check for null. and add the document